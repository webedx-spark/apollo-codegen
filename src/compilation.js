import {
  print,
  visit,
  visitWithTypeInfo,
  typeFromAST,
  getNamedType,
  isAbstractType,
  isEqualType,
  isTypeSubTypeOf,
  doTypesOverlap,
  Kind,
  TypeInfo,
  isType,
  isCompositeType,
  GraphQLScalarType,
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLUnionType,
  GraphQLString,
  GraphQLInt,
  GraphQLFloat,
  GraphQLBoolean,
  GraphQLID,
  GraphQLError
} from 'graphql';

import {
  isTypeProperSuperTypeOf,
  getOperationRootType,
  getFieldDef,
  valueFromValueNode,
  filePathForNode,
  sourceAt,
  withTypenameFieldAddedWhereNeeded,
  isBuiltInScalarType
} from './utilities/graphql';

import {
  join,
  block,
  wrap,
  indent
} from './utilities/printing';

import { 
  flatMap, 
  uniqBy 
} from 'lodash';

import * as sjcl from 'sjcl';

// Parts of this code are adapted from graphql-js

export function compileToIR(schema, document, options = { mergeInFieldsFromFragmentSpreads: true }) {
  if (options.addTypename) {
    document = withTypenameFieldAddedWhereNeeded(schema, document);
  }

  const compiler = new Compiler(schema, document, options);

  const operations = Object.create(null);

  compiler.operations.forEach(operation => {
    operations[operation.name.value] = compiler.compileOperation(operation)
  });

  const fragments = Object.create(null);

  compiler.fragments.forEach(fragment => {
    fragments[fragment.name.value] = compiler.compileFragment(fragment)
  });

  Object.values(operations).forEach(operation => {
    augmentCompiledOperationWithFragments(operation, fragments)
  });

  const typesUsed = compiler.typesUsed;

  return { schema, operations, fragments, typesUsed };
}

export class Compiler {
  constructor(schema, document, options) {
    this.schema = schema;
    this.options = options;

    this.typesUsedSet = new Set();

    this.fragmentMap = Object.create(null);
    this.operations = [];

    for (const definition of document.definitions) {
      switch (definition.kind) {
        case Kind.OPERATION_DEFINITION:
          this.operations.push(definition);
          break;
        case Kind.FRAGMENT_DEFINITION:
          this.fragmentMap[definition.name.value] = definition;
          break;
      }
    }

    this.compiledFragmentMap = Object.create(null);
  }

  addTypeUsed(type) {
    if (this.typesUsedSet.has(type)) return;

    if (type instanceof GraphQLEnumType ||
        type instanceof GraphQLInputObjectType ||
        (type instanceof GraphQLScalarType && !isBuiltInScalarType(type))) {
      this.typesUsedSet.add(type);
    }
    if (type instanceof GraphQLInputObjectType) {
      for (const field of Object.values(type.getFields())) {
        this.addTypeUsed(getNamedType(field.type));
      }
    }
  }

  get typesUsed() {
    return Array.from(this.typesUsedSet);
  }

  fragmentNamed(fragmentName) {
    return this.fragmentMap[fragmentName];
  }

  get fragments() {
    return Object.values(this.fragmentMap);
  }

  compileOperation(operationDefinition) {
    const filePath = filePathForNode(operationDefinition);
    const operationName = operationDefinition.name.value;
    const operationType = operationDefinition.operation;

    const variables = operationDefinition.variableDefinitions.map(node => {
      const name = node.variable.name.value;
      const type = typeFromAST(this.schema, node.type);
      this.addTypeUsed(getNamedType(type));
      return { name, type };
    });

    const source = print(operationDefinition);
    const rootType = getOperationRootType(this.schema, operationDefinition);

    const groupedVisitedFragmentSet = new Map();
    const groupedFieldSet = this.collectFields(rootType, operationDefinition.selectionSet, undefined, groupedVisitedFragmentSet);

    const fragmentsReferencedSet = Object.create(null);
    const { fields } = this.resolveFields(rootType, groupedFieldSet, groupedVisitedFragmentSet, fragmentsReferencedSet);
    const fragmentsReferenced = Object.keys(fragmentsReferencedSet);

    return { filePath, operationName, operationType, rootType, variables, source, fields, fragmentsReferenced };
  }

  compileFragment(fragmentDefinition) {
    const filePath = filePathForNode(fragmentDefinition);
    const fragmentName = fragmentDefinition.name.value;

    const source = print(fragmentDefinition);

    const typeCondition = typeFromAST(this.schema, fragmentDefinition.typeCondition);
    const possibleTypes = this.possibleTypesForType(typeCondition)

    const groupedVisitedFragmentSet = new Map();
    const groupedFieldSet = this.collectFields(typeCondition, fragmentDefinition.selectionSet, undefined, groupedVisitedFragmentSet);

    const fragmentsReferencedSet = Object.create(null);
    const { fields, fragmentSpreads, inlineFragments } = this.resolveFields(typeCondition, groupedFieldSet, groupedVisitedFragmentSet, fragmentsReferencedSet);
    const fragmentsReferenced = Object.keys(fragmentsReferencedSet);

    return { filePath, fragmentName, source, typeCondition, possibleTypes, fields, fragmentSpreads, inlineFragments, fragmentsReferenced };
  }

  collectFields(parentType, selectionSet, groupedFieldSet = Object.create(null), groupedVisitedFragmentSet = new Map()) {
    if (!isCompositeType(parentType)) {
      throw new Error(`parentType should be a composite type, but is "${String(parentType)}"`);
    }

    for (const selection of selectionSet.selections) {
      switch (selection.kind) {
        case Kind.FIELD: {
          const fieldName = selection.name.value;
          const responseName = selection.alias ? selection.alias.value : fieldName;

          const field = getFieldDef(this.schema, parentType, selection);
          if (!field) {
            throw new GraphQLError(`Cannot query field "${fieldName}" on type "${String(parentType)}"`, [selection]);
          }

          if (groupedFieldSet) {
            if (!groupedFieldSet[responseName]) {
              groupedFieldSet[responseName] = [];
            }

            groupedFieldSet[responseName].push([parentType, {
              responseName,
              fieldName,
              args: argumentsFromAST(selection.arguments),
              type: field.type,
              directives: selection.directives,
              selectionSet: selection.selectionSet
            }]);
          }
          break;
        }
        case Kind.INLINE_FRAGMENT: {
          const typeCondition = selection.typeCondition;
          const inlineFragmentType = typeCondition ?
            typeFromAST(this.schema, typeCondition) :
            parentType;

          if (!doTypesOverlap(this.schema, inlineFragmentType, parentType)) continue;
          const effectiveType = parentType instanceof GraphQLObjectType ? parentType : inlineFragmentType;

          this.collectFields(
            effectiveType,
            selection.selectionSet,
            groupedFieldSet,
            groupedVisitedFragmentSet
          );
          break;
        }
        case Kind.FRAGMENT_SPREAD: {
          const fragmentName = selection.name.value;

          const fragment = this.fragmentNamed(fragmentName);
          if (!fragment) throw new GraphQLError(`Cannot find fragment "${fragmentName}"`);

          const typeCondition = fragment.typeCondition;
          const fragmentType = typeFromAST(this.schema, typeCondition)

          if (groupedVisitedFragmentSet) {
            let visitedFragmentSet = groupedVisitedFragmentSet.get(parentType);
            if (!visitedFragmentSet) {
              visitedFragmentSet = {};
              groupedVisitedFragmentSet.set(parentType, visitedFragmentSet);
            }

            if (visitedFragmentSet[fragmentName]) continue;
            visitedFragmentSet[fragmentName] = true;
          }

          if (!doTypesOverlap(this.schema, fragmentType, parentType)) continue;
          const effectiveType = parentType instanceof GraphQLObjectType ? parentType : fragmentType;

          this.collectFields(
            effectiveType,
            fragment.selectionSet,
            this.options.mergeInFieldsFromFragmentSpreads ? groupedFieldSet : null,
            groupedVisitedFragmentSet
          );
          break;
        }
      }
    }

    return groupedFieldSet;
  }

  possibleTypesForType(type) {
    if (isAbstractType(type)) {
      return this.schema.getPossibleTypes(type);
    } else {
      return [type];
    }
  }

  mergeSelectionSets(parentType, fieldSet, groupedVisitedFragmentSet) {
    const groupedFieldSet = Object.create(null);

    for (const [,field] of fieldSet) {
      const selectionSet = field.selectionSet;

      if (selectionSet) {
        this.collectFields(parentType, selectionSet, groupedFieldSet, groupedVisitedFragmentSet);
      }
    }

    return groupedFieldSet;
  }

  resolveFields(parentType, groupedFieldSet, groupedVisitedFragmentSet, fragmentsReferencedSet) {
    const fields = [];

    for (let [responseName, fieldSet] of Object.entries(groupedFieldSet)) {
      fieldSet = fieldSet.filter(([typeCondition,]) => isTypeSubTypeOf(this.schema, parentType, typeCondition));
      if (fieldSet.length < 1) continue;

      const [,firstField] = fieldSet[0];
      const fieldName = firstField.fieldName;
      const args = firstField.args;
      const type = firstField.type;

      let field = { responseName, fieldName, type };

      if (args && args.length > 0) {
        field.args = args;
      }

      const isConditional = fieldSet.some(([,field]) => {
        return field.directives && field.directives.some(directive => {
          const directiveName = directive.name.value;
          return directiveName == 'skip' || directiveName == 'include';
        });
      });

      if (isConditional) {
        field.isConditional = true;
      }

      if (parentType instanceof GraphQLObjectType || parentType instanceof GraphQLInterfaceType) {
        const fieldDef = parentType.getFields()[fieldName];
        if (fieldDef) {
          const description = fieldDef.description;
          if (description) {
            field.description = description
          }

          Object.assign(field, {
            isDeprecated: fieldDef.isDeprecated,
            deprecationReason: fieldDef.deprecationReason,
          });
        }
      }

      const bareType = getNamedType(type);

      this.addTypeUsed(bareType);

      if (isCompositeType(bareType)) {
        const subSelectionGroupedVisitedFragmentSet = new Map();
        const subSelectionGroupedFieldSet = this.mergeSelectionSets(
          bareType,
          fieldSet,
          subSelectionGroupedVisitedFragmentSet
        );

        const { fields, fragmentSpreads, inlineFragments } = this.resolveFields(
          bareType,
          subSelectionGroupedFieldSet,
          subSelectionGroupedVisitedFragmentSet,
          fragmentsReferencedSet
        );
        Object.assign(field, { fields, fragmentSpreads, inlineFragments });
      }

      fields.push(field);
    }

    const fragmentSpreads = this.fragmentSpreadsForParentType(parentType, groupedVisitedFragmentSet);
    const inlineFragments = this.resolveInlineFragments(parentType, groupedFieldSet, groupedVisitedFragmentSet, fragmentsReferencedSet);

    if (fragmentsReferencedSet) {
      Object.assign(fragmentsReferencedSet, ...groupedVisitedFragmentSet.values());

      // TODO: This is a really inefficient way of keeping track of fragments referenced by other fragments
      // We need to either cache compiled fragments or find a way to make resolveFields smarter
      for (let fragmentName of fragmentSpreads) {
        const fragment = this.fragmentNamed(fragmentName);
        if (!fragment) throw new GraphQLError(`Cannot find fragment "${fragmentName}"`);
        const { fragmentsReferenced: fragmentsReferencedFromFragment } = this.compileFragment(fragment);
        for (let fragmentReferenced of fragmentsReferencedFromFragment) {
          fragmentsReferencedSet[fragmentReferenced] = true;
        }
      }
    }

    return { fields, fragmentSpreads, inlineFragments };
  }

  resolveInlineFragments(parentType, groupedFieldSet, groupedVisitedFragmentSet, fragmentsReferencedSet) {
    return this.collectPossibleTypes(parentType, groupedFieldSet, groupedVisitedFragmentSet).map(typeCondition => {
      const { fields, fragmentSpreads } = this.resolveFields(
        typeCondition,
        groupedFieldSet,
        groupedVisitedFragmentSet,
        fragmentsReferencedSet
      );
      const possibleTypes = this.possibleTypesForType(typeCondition)
      return { typeCondition, possibleTypes, fields, fragmentSpreads };
    });
  }

  collectPossibleTypes(parentType, groupedFieldSet, groupedVisitedFragmentSet) {
    if (!isAbstractType(parentType)) return [];

    const possibleTypes = new Set();

    for (const fieldSet of Object.values(groupedFieldSet)) {
      for (const [typeCondition,] of fieldSet) {
        if (this.schema.isPossibleType(parentType, typeCondition)) {
          possibleTypes.add(typeCondition);
        }
      }
    }

    // Also include type conditions for fragment spreads
    if (groupedVisitedFragmentSet) {
      for (const effectiveType of groupedVisitedFragmentSet.keys()) {
        if (this.schema.isPossibleType(parentType, effectiveType)) {
          possibleTypes.add(effectiveType);
        }
      }
    }

    return Array.from(possibleTypes);
  }

  fragmentSpreadsForParentType(parentType, groupedVisitedFragmentSet) {
    if (!groupedVisitedFragmentSet) return [];

    let fragmentSpreads = new Set();

    for (const [effectiveType, visitedFragmentSet] of groupedVisitedFragmentSet) {
      if (!isTypeProperSuperTypeOf(this.schema, effectiveType, parentType)) continue;

      for (const fragmentName of Object.keys(visitedFragmentSet)) {
        fragmentSpreads.add(fragmentName);
      }
    }

    return Array.from(fragmentSpreads);
  }
}

function augmentCompiledOperationWithFragments(compiledOperation, compiledFragments) {
  const operationAndFragments = operationAndRelatedFragments(compiledOperation, compiledFragments);
  compiledOperation.sourceWithFragments = operationAndFragments.map(operationOrFragment => { 
    return operationOrFragment.source; 
  }).join('\n');
  const idBits = sjcl.hash.sha256.hash(compiledOperation.sourceWithFragments);
  compiledOperation.operationId = sjcl.codec.hex.fromBits(idBits);
}

function operationAndRelatedFragments(compiledOperationOrFragment, allCompiledFragments) {
  let result = flatMap(compiledOperationOrFragment.fragmentsReferenced, (fragmentName) => {
    return operationAndRelatedFragments(allCompiledFragments[fragmentName], allCompiledFragments);
  });
  result.unshift(compiledOperationOrFragment);
  result = uniqBy(result, (compiledOperationOrFragment) => {
    return compiledOperationOrFragment.fragmentName;
  });
  result = result.sort((a, b) => {
    return a.fragmentName > b.fragmentName;
  });
  return result;
}

function argumentsFromAST(args) {
  return args && args.map(arg => {
    return { name: arg.name.value, value: valueFromValueNode(arg.value) };
  });
}

export function printIR({ fields, inlineFragments, fragmentSpreads }) {
  return fields && wrap('<', join(fragmentSpreads, ', '), '> ')
    + block(fields.map(field =>
      `${field.name}: ${String(field.type)}` + wrap(' ', printIR(field))
    ).concat(inlineFragments && inlineFragments.map(inlineFragment =>
      `${String(inlineFragment.typeCondition)}` + wrap(' ', printIR(inlineFragment)))));
}
