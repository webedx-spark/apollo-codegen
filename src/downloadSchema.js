// Based on https://facebook.github.io/relay/docs/guides-babel-plugin.html#using-other-graphql-implementations

import * as fetch from 'node-fetch';
import * as fs from 'fs';
import * as path from 'path';
import * as https from 'https';

import {
  buildClientSchema,
  introspectionQuery,
  printSchema,
} from 'graphql/utilities';

import { ToolError } from './errors'

const defaultHeaders = {
  'Accept': 'application/json',
  'Content-Type': 'application/json'
};

export default async function downloadSchema(url, outputPath, additionalHeaders, insecure, method) {
  const headers = Object.assign(defaultHeaders, additionalHeaders);
  const agent = insecure ? new https.Agent({ rejectUnauthorized: false }) : null;

  let result;
  try {
    const response = await fetch(url, {
      method: method,
      headers: headers,
      body: JSON.stringify({ 'query': introspectionQuery }),
      agent,
    });

    result = await response.json();
  } catch (error) {
    throw new ToolError(`Error while fetching introspection query result: ${error.message}`);
  }

  if (result.errors) {
    throw new ToolError(`Errors in introspection query result: ${result.errors}`);
  }

  const schemaData = result;
  if (!schemaData.data) {
    throw new ToolError(`No introspection query result data found, server responded with: ${JSON.stringify(result)}`);
  }

  fs.writeFileSync(outputPath, JSON.stringify(schemaData, null, 2));
}
