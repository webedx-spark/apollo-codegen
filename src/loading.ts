import * as path from 'path'
import * as fs from 'fs'
import * as mkdirp from 'mkdirp'

import {
  buildClientSchema,
  Source,
  concatAST,
  parse,
  DocumentNode,
  GraphQLSchema
} from 'graphql';

import { ToolError, logError } from './errors'

export function loadSchema(schemaPath: string): GraphQLSchema {
  if (!fs.existsSync(schemaPath)) {
    throw new ToolError(`Cannot find GraphQL schema file: ${schemaPath}`);
  }
  const schemaData = require(schemaPath);

  if (!schemaData.data && !schemaData.__schema) {
    throw new ToolError('GraphQL schema file should contain a valid GraphQL introspection query result');
  }
  return buildClientSchema((schemaData.data) ? schemaData.data : schemaData);
}

function extractDocumentFromJavascript(content: string): string | null {
  const re = /gql`([^`]*)`/g;
  let match
  const matches = []

  while(match = re.exec(content)) {
    const doc = match[1]
      .replace(/\${[^}]*}/g, '')

    matches.push(doc)
  }

  const doc = matches.join('\n')
  return doc.length ? doc : null;
}

export function loadAndMergeQueryDocuments(inputPaths: string[]): DocumentNode {
  const sources = inputPaths.map(inputPath => {
    const body = fs.readFileSync(inputPath, 'utf8');
    if (!body) {
      return null;
    }

    if (inputPath.endsWith('.jsx') || inputPath.endsWith('.js')
      || inputPath.endsWith('.tsx') || inputPath.endsWith('.ts')
    ) {
      const doc = extractDocumentFromJavascript(body.toString());
      if (doc) {
        if (doc.trim().indexOf('mutation') === 0) {
          return null;
        }

        return new Source(doc, inputPath);
      } else {
        return null;
      }
    }

    if (body.trim().indexOf('mutation') === 0) {
      return null;
    }

    return new Source(body, inputPath);
  }).filter(source => source);

  return concatAST((sources as Source[]).map(source => parse(source)));
}
