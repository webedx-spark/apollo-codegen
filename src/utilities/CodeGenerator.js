import {
  join,
  wrap,
} from './printing';

export default class CodeGenerator {
  constructor(context) {
    this.context = context;
    
    this.scopeStack = [];
    this.queue = [];

    this.indentWidth = 2;
    this.indentLevel = 0;

    this.output = '';
  }

  pushScope(scope) {
    this.scopeStack.push(scope);
  }

  popScope() {
    return this.scopeStack.pop();
  }

  print(maybeString) {
    if (maybeString) {
      this.output += maybeString;
    }
  }

  printNewline() {
    if (this.output) {
      this.print('\n');
      this.startOfIndentLevel = false;
    }
  }

  printNewlineIfNeeded() {
    if (!this.startOfIndentLevel) {
      this.printNewline();
    }
  }

  printOnNewline(maybeString) {
    if (maybeString) {
      this.printNewline();
      this.printIndent();
      this.print(maybeString);
    }
  }

  printIndent() {
    const indentation = ' '.repeat(this.indentLevel * this.indentWidth);
    this.output += indentation;
  }

  withIndent(closure) {
    if (!closure) return;

    this.indentLevel++;
    this.startOfIndentLevel = true;
    closure();
    this.indentLevel--;
  }

  withinBlock(closure, open = ' {', close = '}', closeOnNewLine = true) {
    this.print(open);
    this.withIndent(closure);
    if (closeOnNewLine) {
      this.printOnNewline(close);
    } else {
      this.print(close);
    }
  }

  queueBlock(closure, open = ' {', close = '}') {
    this.queue.push(closure);
  }

  flushQueued(closure) {
    while (this.queue.length > 0) {
      this.queue.pop()();
    }
  }
}
