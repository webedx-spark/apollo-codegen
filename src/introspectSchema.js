import * as fs from 'fs';

import { buildASTSchema, graphql, parse } from 'graphql';
import { introspectionQuery } from 'graphql/utilities';

import { ToolError } from './errors'

export async function introspect(schemaContents) {
  const schema = buildASTSchema(parse(schemaContents));
  return await graphql(schema, introspectionQuery);
}

export default async function introspectSchema(schemaPath, outputPath) {
  if (!fs.existsSync(schemaPath)) {
    throw new ToolError(`Cannot find GraphQL schema file: ${schemaPath}`);
  }

  const schemaContents = fs.readFileSync(schemaPath).toString();
  const result = await introspect(schemaContents);

  if (result.errors) {
    throw new ToolError(`Errors in introspection query result: ${result.errors}`);
  }

  fs.writeFileSync(outputPath, JSON.stringify(result, null, 2));
}
