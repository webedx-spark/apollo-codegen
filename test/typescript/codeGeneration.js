import { stripIndent } from 'common-tags';

import {
  parse,
  isType,
  GraphQLID,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull
} from 'graphql';

import {
  generateSource
} from '../../src/typescript/codeGeneration';

import { loadSchema } from '../../src/loading';
const swapiSchema = loadSchema(require.resolve('../starwars/schema.json'));
const miscSchema = loadSchema(require.resolve('../misc/schema.json'));

import CodeGenerator from '../../src/utilities/CodeGenerator';

import { compileToIR } from '../../src/compilation';

describe('TypeScript code generation', function() {
  let generator;
  let compileFromSource;
  let addFragment;

  function setup(schema) {
    const context = {
      schema: schema,
      operations: {},
      fragments: {},
      typesUsed: {}
    }

    generator = new CodeGenerator(context);

    compileFromSource = (source) => {
      const document = parse(source);
      const context = compileToIR(schema, document, {
        mergeInFieldsFromFragmentSpreads: true,
        addTypename: true
      });
      generator.context = context;
      return context;
    };

    addFragment = (fragment) => {
      generator.context.fragments[fragment.fragmentName] = fragment;
    };

    return { generator, compileFromSource, addFragment };
  }

  describe('#generateSource()', function() {
    test(`should generate simple query operations`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroName {
          hero {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate simple query operations including input variables`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroName($episode: Episode) {
          hero(episode: $episode) {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate simple nested query operations including input variables`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              name
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate fragmented query operations`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndFriendsNames {
          hero {
            name
            ...heroFriends
          }
        }

        fragment heroFriends on Character {
          friends {
            name
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate query operations with inline fragments`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        query HeroAndDetails {
          hero {
            name
            ...HeroDetails
          }
        }

        fragment HeroDetails on Character {
          ... on Droid {
            primaryFunction
          }
          ... on Human {
            height
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate mutation operations with complex input types`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        mutation ReviewMovie($episode: Episode, $review: ReviewInput) {
          createReview(episode: $episode, review: $review) {
            stars
            commentary
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test(`should generate correct list with custom fragment`, function() {
      const { compileFromSource } = setup(swapiSchema);
      const context = compileFromSource(`
        fragment Friend on Character {
          name
        }

        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              ...Friend
            }
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test('should handle single line comments', () => {
      const { compileFromSource } = setup(miscSchema);
      const context = compileFromSource(`
        query CustomScalar {
          commentTest {
            singleLine
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });

    test('should handle multi-line comments', () => {
      const { compileFromSource } = setup(miscSchema);
      const context = compileFromSource(`
        query CustomScalar {
          commentTest {
            multiLine
          }
        }
      `);

      const source = generateSource(context);
      expect(source).toMatchSnapshot();
    });
  });
});
