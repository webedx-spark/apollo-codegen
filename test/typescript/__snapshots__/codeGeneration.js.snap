// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript code generation #generateSource() should generate correct list with custom fragment 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroAndFriendsNamesQueryVariables = {
  episode: Episode | null,
};

export type HeroAndFriendsNamesQuery = {
  hero: ( {
      __typename: \\"Human\\",
      // The name of the character
      name: string,
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    } | {
      __typename: \\"Droid\\",
      // The name of the character
      name: string,
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    }
  ) | null,
};

export type FriendFragment = ( {
      __typename: \\"Human\\",
      // The name of the character
      name: string,
    } | {
      __typename: \\"Droid\\",
      // The name of the character
      name: string,
    }
  );
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate fragmented query operations 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

export type HeroAndFriendsNamesQuery = {
  hero: ( {
      __typename: \\"Human\\",
      // The name of the character
      name: string,
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    } | {
      __typename: \\"Droid\\",
      // The name of the character
      name: string,
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    }
  ) | null,
};

export type HeroFriendsFragment = ( {
      __typename: \\"Human\\",
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    } | {
      __typename: \\"Droid\\",
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    }
  );
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate mutation operations with complex input types 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type ReviewInput = {
  // 0-5 stars
  stars: number,
  // Comment about the movie, optional
  commentary: string | null,
  // Favorite color, optional
  favorite_color: ColorInput | null,
};

export type ColorInput = {
  red: number,
  green: number,
  blue: number,
};

export type ReviewMovieMutationVariables = {
  episode: Episode | null,
  review: ReviewInput | null,
};

export type ReviewMovieMutation = {
  createReview:  {
    __typename: string,
    // The number of stars this review gave, 1-5
    stars: number,
    // Comment about the movie
    commentary: string | null,
  } | null,
};
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate query operations with inline fragments 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

export type HeroAndDetailsQuery = {
  hero: ( {
      __typename: \\"Human\\",
      // What this human calls themselves
      name: string,
      // Height in the preferred unit, default is meters
      height: number | null,
    } | {
      __typename: \\"Droid\\",
      // What others call this droid
      name: string,
      // This droid's primary function
      primaryFunction: string | null,
    }
  ) | null,
};

export type HeroDetailsFragment = ( {
      __typename: \\"Human\\",
      // Height in the preferred unit, default is meters
      height: number | null,
    } | {
      __typename: \\"Droid\\",
      // This droid's primary function
      primaryFunction: string | null,
    }
  );
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate simple nested query operations including input variables 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroAndFriendsNamesQueryVariables = {
  episode: Episode | null,
};

export type HeroAndFriendsNamesQuery = {
  hero: ( {
      __typename: \\"Human\\",
      // The name of the character
      name: string,
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    } | {
      __typename: \\"Droid\\",
      // The name of the character
      name: string,
      // The friends of the character, or an empty list if they have none
      friends:  Array<( {
          __typename: \\"Human\\",
          // The name of the character
          name: string,
        } | {
          __typename: \\"Droid\\",
          // The name of the character
          name: string,
        }
      ) > | null,
    }
  ) | null,
};
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate simple query operations 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

export type HeroNameQuery = {
  hero: ( {
      __typename: \\"Human\\",
      // The name of the character
      name: string,
    } | {
      __typename: \\"Droid\\",
      // The name of the character
      name: string,
    }
  ) | null,
};
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should generate simple query operations including input variables 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

// The episodes in the Star Wars trilogy
export type Episode =
  \\"NEWHOPE\\" | // Star Wars Episode IV: A New Hope, released in 1977.
  \\"EMPIRE\\" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
  \\"JEDI\\"; // Star Wars Episode VI: Return of the Jedi, released in 1983.


export type HeroNameQueryVariables = {
  episode: Episode | null,
};

export type HeroNameQuery = {
  hero: ( {
      __typename: \\"Human\\",
      // The name of the character
      name: string,
    } | {
      __typename: \\"Droid\\",
      // The name of the character
      name: string,
    }
  ) | null,
};
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should handle multi-line comments 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

export type CustomScalarQuery = {
  commentTest:  {
    __typename: string,
    // This is a multi-line
    // comment.
    multiLine: string | null,
  } | null,
};
/* tslint:enable */
"
`;

exports[`TypeScript code generation #generateSource() should handle single line comments 1`] = `
"/* tslint:disable */
//  This file was automatically generated and should not be edited.

export type CustomScalarQuery = {
  commentTest:  {
    __typename: string,
    // This is a single-line comment
    singleLine: string | null,
  } | null,
};
/* tslint:enable */
"
`;
